### ISSUE TITLE
memcopy and join have launch configuration issues

### ISSUE THREAD

**Issue by grzegorz-gutowski** [arrayfire/arrayfire#1513]
Hello,
I'm running the following code that exploits wrap/unwrap to mix/unmix channels of an rgb image.

``` cpp
#include <iostream>

#include <arrayfire.h>

af::array mix_chans(const af::array& a, const af::array& b, const af::array& c) {
    unsigned d0 = a.dims(0);
    unsigned d1 = a.dims(1);
    auto a_ = af::moddims(a, 1, d0*d1);
    auto b_ = af::moddims(b, 1, d0*d1);
    auto c_ = af::moddims(c, 1, d0*d1);
    auto u = af::join(0, a_, b_, c_); 
    return af::wrap(u, d0*3, d1, 3, 1, 3, 1, 0, 0);
}
void unmix_chans(const af::array& abc, af::array& a, af::array& b, af::array& c) {
    unsigned d0 = abc.dims(0)/3;
    unsigned d1 = abc.dims(1);
    auto abc_ = af::flat(abc);
    auto u = af::unwrap(abc_, 3, 1, 3, 1, 0, 0);
    a = af::moddims(u(0, af::span, af::span, af::span), d0, d1);
    b = af::moddims(u(1, af::span, af::span, af::span), d0, d1);
    c = af::moddims(u(2, af::span, af::span, af::span), d0, d1);
}

int main() {
    int w = 1920;
    int h = 1080;
    af::array r = af::randu(w, h);
    af::array g = af::randu(w, h);
    af::array b = af::randu(w, h);
    af::array rgb = mix_chans(r, g, b);
    af::array r_,g_,b_;
    unmix_chans(rgb, r_, g_, b_);
    std::cerr << af::max<double>(af::abs(r - r_)) << " " <<
                 af::max<double>(af::abs(g - g_)) << " " <<
                 af::max<double>(af::abs(b - b_)) << std::endl;
}
```

I get the following CUDA/AF Internal error:

```
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Internal error:998):
In function memcopy
In file src/backend/cuda/kernel/memcopy.hpp:85
CUDA Error (9): invalid configuration argument


In function af::array af::moddims(const af::array&, unsigned int, const dim_t*)
In file src/api/cpp/data.cpp:265
Aborted
```

When dimensions of the image are halved then everything runs fine.
I'm using ArrayFire v3.3.2 (CUDA, 64-bit Linux, build f65dd97).

**Comment by pavanky** [comment]
What GPU are you using?

**Comment by grzegorz-gutowski** [comment]
GTX950 and GTX980 - same results on both.

**Comment by grzegorz-gutowski** [comment]
Today, I've tried the same code with github head:

    ArrayFire v3.4.0 (CUDA, 64-bit Linux, build 072d507)
    Platform: CUDA Toolkit 7.5, Driver: 361.42
    [0] GeForce GTX 980, 4096 MB, CUDA Compute 5.2

and there is no runtime error, but I get wrong results: after chanel mixing/unmixing the arrays differ! Small arrays are still ok.

**Comment by pavanky** [comment]
@grzegorz-gutowski We haven't fixed the original bug..  I just checked same commit on my GTX 950 and still get the original error.

Are you sure you are running the same code ?

**Comment by pavanky** [comment]
@grzegorz-gutowski BTW, this is a simpler way to achieve what you are trying to do and also has the benefit of not erroring out.

```cpp
af::array mix_chans(const af::array& a, const af::array& b, const af::array& c) {
    unsigned d0 = a.dims(0);
    unsigned d1 = a.dims(1);
    auto a_ = af::moddims(a, 1, d0, d1);
    auto b_ = af::moddims(b, 1, d0, d1);
    auto c_ = af::moddims(c, 1, d0, d1);
    auto u = af::join(0, a_, b_, c_);
    return af::moddims(u, d0*3, d1);
}
void unmix_chans(const af::array& abc, af::array& a, af::array& b, af::array& c) {
    unsigned d0 = abc.dims(0)/3;
    unsigned d1 = abc.dims(1);
    auto abc_ = af::moddims(abc, 3, d0, d1);
    a = af::moddims(abc_(0, af::span, af::span, af::span), d0, d1);
    b = af::moddims(abc_(1, af::span, af::span, af::span), d0, d1);
    c = af::moddims(abc_(2, af::span, af::span, af::span), d0, d1);
}
```

**Comment by grzegorz-gutowski** [comment]
@pavanky Thanks for the solution: it's cleaner and works great for me.
I confirm that the behavior of builds 072d507, and f65dd97 differs with the same example code.

**Comment by shehzan10** [comment]
This issue is a part of #133. In this particular example, the `memcopy` and `join` kernels have launch configuration issues. Will have a fix for these two out soon.

**Comment by shehzan10** [comment]
Pushing this back to v3.5.0 so that it can use a better, more robust solution.

**Comment by pavanky** [comment]
Fixed in v3.5

**Comment by pavanky** [comment]
Fixed in v3.5
